Fix /givecashall debug-command

skruw opened the pull request:
Closes #15207

Hi there. This is my second PR. Hope my solution is correct :). Tested it in all four mods and it worked for me. 

- Fix Order-Validaiton for the order "DevGiveCashAll"
- Fix Game.Debug-Outputfor the order "DevGiveCashAll"

Smittytron commented:
Can you squash the fix-up commit?

skruw commented:
I'll look into it.

abcdefg30 commented on OpenRA.Mods.Common/Commands/DevCommands.cs:
```diff
@@ -156,6 +134,23 @@ public void InvokeCommand(string name, string arg)
 
+		void GiveCash(bool toAll, string arg)
+		{
+			string orderString = "";
```
This can be shortened to
`var orderString = toAll ? "DevGiveCashAll" : "DevGiveCash";`

skruw commented:
Thank you for your suggestion. I changed the line.

pchote commented:
This is indeed what I had in mind broadly speaking.  There are a couple of code style nits i'll ask you to change when I have time to look at this properly - please ping me again here if I don't review this by the end of the weekend.

pchote commented on OpenRA.Mods.Common/Traits/Player/DeveloperMode.cs:
```diff
@@ -161,10 +162,28 @@ public void ResolveOrder(Actor self, Order order)
 					var amount = order.ExtraData != 0 ? (int)order.ExtraData : info.Cash;
+
+					var playerName = order.Player.PlayerName;
+					Game.Debug("Giving {0} credits to player {1}.", order.ExtraData == 0 ? "cheat default" : amount.ToString(CultureInfo.InvariantCulture), playerName);
```
This can simplify down to `Game.Debug("Giving {0} credits to player {1}.", amount, order.Player.PlayerName);` (and then remove the separate `playerName` variable).

pchote reviewed (commented):
A few small comments, then LGTM:

skruw commented:
Done.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Player/DeveloperMode.cs:
```diff
@@ -161,10 +162,27 @@ public void ResolveOrder(Actor self, Order order)
+
+					Game.Debug("Giving {0} credits to all players.", amount);
+					foreach (var player in receivingPlayers)
+					{
```
Please omit braces around single line statements:
```c#
foreach (var player in receivingPlayers)
	player.PlayerActor.Trait<PlayerResources>().ChangeCash(amount);
```

abcdefg30 commented on OpenRA.Mods.Common/Traits/Player/DeveloperMode.cs:
```diff
@@ -161,10 +162,27 @@ public void ResolveOrder(Actor self, Order order)
+				case "DevGiveCashAll":
+				{
+					var amount = order.ExtraData != 0 ? (int)order.ExtraData : info.Cash;
+					var receivingPlayers = self.World.Players.Where(p => p.Playable);
```
Is there a reason you changed `!NonCombatant` to `Playable`?

pchote replied:
IMO this change makes sense. It gives cash to all the players who have slots in the lobby, but omits scripted players who may otherwise break if they unexpectedly receive or lose cash.

abcdefg30 replied:
This really hurts now when testing campaign AIs...

abcdefg30 reviewed (commented):
Sorry for taking so long. A few final nits/questions before we finally merge this.

skruw committed:
Fix givecashall debug command

Issuing an order to another client causes a validation error in
ValidateOrder.OrderValidation(...). To fix this /givecashall will now
be issued as an order to the client that introduced the command. This
client will then resolve the command and give the cash to all playable
parties.

pchote commented:
I've pushed over this with a fix for the braces and a tweak to the chat output.
Also rebased on latest bleed.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/25cdd65e41eb5f62ef28ccfaa53438ca77a9e109...948e5e0421ee515b0c5cd0297cd35da2b2495983)
